dm/syncer/data_validator.go

```diff
@@ -656,40 +681,32 @@ func (v *DataValidator) processRowsEvent(header *replication.EventHeader, ev *re
 		return nil
 	}
 
-	table, err := v.genValidateTableInfo(sourceTable)
-	if err != nil {
-		if schema.IsTableNotExists(err) {
-			// not a table need to sync
-			return nil
-		} else if err == errNoPrimaryKey {
-			v.tableStatus[fullTableName] = &tableValidateStatus{
-				source:  *sourceTable,
-				target:  *table.Target,
-				stage:   pb.Stage_Stopped,
-				message: tableWithoutPrimaryKeyMsg,
-			}
-			return nil
-		}
-		return terror.Annotate(err, "failed to get table info")
+	fullTableName := sourceTable.String()
+	state, ok := v.tableStatus[fullTableName]
+	if ok && state.stage == pb.Stage_Stopped {
+		return nil
 	}
 
-	if len(table.Info.Columns) < int(ev.ColumnCount) {
-		v.tableStatus[fullTableName] = &tableValidateStatus{
-			source:  *sourceTable,
-			target:  *table.Target,
-			stage:   pb.Stage_Stopped,
-			message: moreColumnInBinlogMsg,
-		}
-		return nil
+	validateTbl, err := v.genValidateTableInfo(sourceTable, int(ev.ColumnCount))
+	if err != nil {
+		return terror.Annotate(err, "failed to get table info")
 	}
 
-	if _, ok := v.tableStatus[fullTableName]; !ok {
-		v.tableStatus[fullTableName] = &tableValidateStatus{
+	targetTable := validateTbl.targetTable
+	if state == nil {
+		state = &tableValidateStatus{
 			source: *sourceTable,
-			target: *table.Target,
+			target: *targetTable,
 			stage:  pb.Stage_Running,
 		}
+		v.tableStatus[fullTableName] = state
 	}
+	if validateTbl.message != "" {
+		state.stopped(validateTbl.message)
```

Comment:
this worker will stop when column count mismatch, do we need to add a log here?

Reply:
fix in https://github.com/pingcap/tiflow/pull/5254/commits/7efda04e17ac61c3bde2f5d45a3be5f53e10cb11